File: frontend/src/api/optimize.ts
import apiClient from './client'

export type OptimizeSupportLevel = 'full' | 'partial' | 'static_only' | 'unsupported'
export type OptimizeSource = 'manual' | 'slowlog' | 'fingerprint'

export interface OptimizeFinding {
  severity: 'critical' | 'warning' | 'info' | 'ok'
  code: string
  title: string
  detail: string
  evidence: string
  confidence: number
}

export interface OptimizeRecommendation {
  priority: number
  type: string
  title: string
  action: string
  reason: string
  risk: string
  confidence: number
}

export interface OptimizeAnalyzeResponse {
  supported: boolean
  support_level: OptimizeSupportLevel
  engine: string
  source: OptimizeSource
  risk_score: number
  summary: string
  findings: OptimizeFinding[]
  recommendations: OptimizeRecommendation[]
  plan: {
    format: string
    summary: Record<string, any>
    operators: Record<string, any>[]
  }
  metadata: {
    tables: string[]
    indexes: Record<string, any>[]
    statistics: Record<string, any>[]
    slowlog: Record<string, any>
  }
  raw: any
  sql: string
  msg: string
}

export const optimizeApi = {
  analyze: (data: {
    log_id?: number
    fingerprint?: string
    instance_id?: number
    db_name?: string
    sql?: string
  }) => apiClient.post<OptimizeAnalyzeResponse>('/optimize/analyze/', data).then(r => r.data),
}

File: backend/app/schemas/optimize.py
"""SQL optimization v2 schemas."""

from typing import Any, Literal

from pydantic import BaseModel, Field, model_validator

SupportLevel = Literal["full", "partial", "static_only", "unsupported"]
OptimizeSource = Literal["manual", "slowlog", "fingerprint"]


class OptimizeAnalyzeRequest(BaseModel):
    """Unified SQL optimization analysis request."""

    log_id: int | None = None
    fingerprint: str | None = None
    instance_id: int | None = None
    db_name: str = ""
    sql: str = ""

    @model_validator(mode="after")
    def validate_source(self) -> "OptimizeAnalyzeRequest":
        if self.log_id or self.fingerprint:
            return self
        if not self.instance_id or not self.sql.strip():
            raise ValueError("必须提供 log_id、fingerprint，或 instance_id + sql")
        return self


class OptimizeFinding(BaseModel):
    severity: Literal["critical", "warning", "info", "ok"] = "info"
    code: str
    title: str
    detail: str
    evidence: str = ""
    confidence: float = Field(default=0.7, ge=0, le=1)


class OptimizeRecommendation(BaseModel):
    priority: int = Field(default=99, ge=1)
    type: str = "general"
    title: str
    action: str
    reason: str = ""
    risk: str = ""
    confidence: float = Field(default=0.7, ge=0, le=1)


class OptimizePlan(BaseModel):
    format: str = ""
    summary: dict[str, Any] = Field(default_factory=dict)
    operators: list[dict[str, Any]] = Field(default_factory=list)


class OptimizeMetadata(BaseModel):
    tables: list[str] = Field(default_factory=list)
    indexes: list[dict[str, Any]] = Field(default_factory=list)
    statistics: list[dict[str, Any]] = Field(default_factory=list)
    slowlog: dict[str, Any] = Field(default_factory=dict)


class OptimizeAnalyzeResponse(BaseModel):
    supported: bool = True
    support_level: SupportLevel = "static_only"
    engine: str = ""
    source: OptimizeSource = "manual"
    risk_score: int = Field(default=0, ge=0, le=100)
    summary: str = ""
    findings: list[OptimizeFinding] = Field(default_factory=list)
    recommendations: list[OptimizeRecommendation] = Field(default_factory=list)
    plan: OptimizePlan = Field(default_factory=OptimizePlan)
    metadata: OptimizeMetadata = Field(default_factory=OptimizeMetadata)
    raw: Any = None
    sql: str = ""
    msg: str = ""

